### PR TITLE
Estimate bandwidth in connection and don't limit by REMBs

### PR DESCRIPTION
**Pull request by lodoyun**
**Description**
This PR fundamentally changes the way `maxVideoBW` works:
1. The bitrate estimation for the receiver is no longer done at the mediaStream level but at the connection level.
2. We no longer limit the publishers bitrate by using REMBs but by modifying the encoder sender parameters.

[] It needs and includes Unit Tests

**Changes in Client or Server public APIs**

The APIs are the same but `maxVideoBW` in the publisher now has a slightly different meaning, particularly in simulcast. Now that bandwidth will limit the top layer instead of limiting the whole stream (the addition of all the layers bitrates). When not in simulcast the result is more or less the same since we only have one layer.

If more control over layer bitrate is needed, apps should use `updateSimulcastLayersBitrate` to set the bitrate for each layer.

[] It includes documentation for these changes in `/doc`.

### CODE REVIEW

File: erizo/src/erizo/rtp/BandwidthEstimationHandler.cpp [hunk 1]
```diff
@@ -57,32 +58,25 @@ void BandwidthEstimationHandler::disable() {
 }
 
 void BandwidthEstimationHandler::notifyUpdate() {
+  ELOG_DEBUG("NotifyUPDATE");
```
Comment (jcague) [inline erizo/src/erizo/rtp/BandwidthEstimationHandler.cpp:61]: should we use a more descriptive message?

File: erizo/src/erizo/rtp/BandwidthEstimationHandler.cpp [hunk 2]
```diff
@@ -229,19 +223,39 @@ void BandwidthEstimationHandler::pickEstimator() {
 }
 
 void BandwidthEstimationHandler::sendREMBPacket() {
+  sink_ssrc_ = 0;
```
Comment (jcague) [inline erizo/src/erizo/rtp/BandwidthEstimationHandler.cpp:226]: why do we need to use a class variable?

File: erizo/src/erizo/rtp/BandwidthEstimationHandler.cpp [hunk 3]
```diff
@@ -229,19 +223,39 @@ void BandwidthEstimationHandler::pickEstimator() {
...
+    ELOG_DEBUG("MediaStream %s, publisher %u, sink %u, source %u", media_stream->getId().c_str(),
+    media_stream->isPublisher(), media_stream->getVideoSinkSSRC(), media_stream->getVideoSourceSSRC());
+    if (media_stream->isReady() && media_stream->isPublisher()) {
+      sink_ssrc_ = media_stream->getVideoSinkSSRC();
+    }
+    source_ssrcs_.push_back(media_stream->getVideoSourceSSRC());
+  });
+
+  if (sink_ssrc_ == 0) {
+    ELOG_WARN("No SSRC available to send REMB");
```
Comment (jcague) [inline erizo/src/erizo/rtp/BandwidthEstimationHandler.cpp:239]: I wonder if there could be cases where we should not notify with a warning here.

File: erizo/src/erizo/rtp/BandwidthEstimationHandler.cpp [hunk 4]
```diff
@@ -252,6 +266,7 @@ void BandwidthEstimationHandler::sendREMBPacket() {
 
 void BandwidthEstimationHandler::OnReceiveBitrateChanged(const std::vector<uint32_t>& ssrcs,
                                      uint32_t bitrate) {
+  ELOG_WARN("Onreceive bitrate %lu", bitrate);
```
Comment (jcague) [inline erizo/src/erizo/rtp/BandwidthEstimationHandler.cpp:269]: consider using ELOG_DEBUG here

File: erizo_controller/erizoClient/src/Stream.js [hunk 5]
```diff
@@ -68,6 +77,77 @@ const Stream = (altConnectionHelpers, specInput) => {
     that.local = true;
   }
 
+  const setMaxVideoBW = (maxVideoBW) => {
+    if (that.local) {
+      const translated = (maxVideoBW * 1000 * 0.90) - (50 * 40 * 8);
```
Comment (jcague) [inline erizo_controller/erizoClient/src/Stream.js:82]: I think we could add a reference here to know where this formula comes from

File: erizo_controller/erizoClient/src/Stream.js [hunk 6]
```diff
@@ -68,6 +77,77 @@ const Stream = (altConnectionHelpers, specInput) => {
...
+      videoSenderLicodeParameters[index] = {};
+    }
+    if (that.maxVideoBW) {
+      log.warning('Setting maxVideoBW', that.maxVideoBW);
+      videoSenderLicodeParameters[layersToConfigure - 1].maxBitrate = that.maxVideoBW;
+    }
+  };
+
+  const configureVideoStream = (options) => {
+    log.warning('configureVideoStream', options);
```
Comment (jcague) [inline erizo_controller/erizoClient/src/Stream.js:140]: consider `log.debug` instead

File: erizo_controller/erizoClient/src/Stream.js [hunk 7]
```diff
@@ -68,6 +77,77 @@ const Stream = (altConnectionHelpers, specInput) => {
...
+    if (!simulcastConfig) {
+      videoSenderLicodeParameters[0] = { maxBitrate: that.maxVideoBW }; // No simulcast
+      return;
+    }
+    const layersToConfigure = simulcastConfig.numSpatialLayers;
+    for (let index = 0; index < layersToConfigure; index += 1) {
+      videoSenderLicodeParameters[index] = {};
+    }
+    if (that.maxVideoBW) {
+      log.warning('Setting maxVideoBW', that.maxVideoBW);
```
Comment (jcague) [inline erizo_controller/erizoClient/src/Stream.js:134]: consider using `log.debug` here

File: erizo_controller/erizoClient/src/webrtc-stacks/BaseStack.js [hunk 8]
```diff
@@ -14,7 +14,7 @@ const BaseStack = (specInput) => {
   };
   that.getNegotiationLogs = () => logs.reduce((a, b) => `${a}'\n'${b}`);
 
-  log.debug(`message: Starting Base stack, spec: ${JSON.stringify(specBase)}`);
+  log.warning(`message: Starting Base stack, spec: ${JSON.stringify(specBase)}`);
```
Comment (jcague) [inline erizo_controller/erizoClient/src/webrtc-stacks/BaseStack.js:17]: consider using `log.debug` here again

File: erizo/src/erizo/rtp/BandwidthEstimationHandler.cpp [hunk 9]
```diff
@@ -229,19 +223,39 @@ void BandwidthEstimationHandler::pickEstimator() {
...
+  uint32_t capped_bitrate = bitrate_;
+  ELOG_DEBUG("Bitrates min(%u) = %u", bitrate_, capped_bitrate);
   remb_packet_.setREMBBitRate(capped_bitrate);
-  remb_packet_.setREMBNumSSRC(1);
-  remb_packet_.setREMBFeedSSRC(0, stream_->getVideoSourceSSRC());
+  remb_packet_.setREMBNumSSRC(source_ssrcs_.size());
+
+  for (std::size_t i = 0; i < source_ssrcs_.size(); i++) {
+    ELOG_DEBUG("Setting REMBFeedSSRC %u to ssrc %u, size %u", i, source_ssrcs_[i], source_ssrcs_.size());
+    remb_packet_.setREMBFeedSSRC(i, source_ssrcs_[i]);
```
Comment (jcague) [inline erizo/src/erizo/rtp/BandwidthEstimationHandler.cpp:256]: 💯

File: erizo_controller/erizoClient/src/Stream.js [hunk 10]
```diff
@@ -130,7 +210,26 @@ const Stream = (altConnectionHelpers, specInput) => {
...
 
-  that.addPC = (pc, p2pKey = undefined) => {
+  that.getMaxVideoBW = () => that.maxVideoBW;
+  that.hasSimulcast = () => Object.keys(videoSenderLicodeParameters).length > 1;
+
+  that.generateEncoderParameters = () => {
+    const nativeSenderParameters = [];
+    const requestedLayers = Object.keys(videoSenderLicodeParameters).length ||
+      defaultSimulcastSpatialLayers;
+    const isScreenshare = that.hasScreen();
```
Comment (jcague) [inline erizo_controller/erizoClient/src/Stream.js:220]: 💯

File: erizo_controller/erizoClient/src/Stream.js [hunk 11]
```diff
@@ -68,6 +77,77 @@ const Stream = (altConnectionHelpers, specInput) => {
...
   }
 
+  const setMaxVideoBW = (maxVideoBW) => {
+    if (that.local) {
+      const translated = (maxVideoBW * 1000 * 0.90) - (50 * 40 * 8);
+      log.info(`message: Setting maxVideoBW, streamId: ${that.getID()}, maxVideoBW: ${maxVideoBW}, translated: ${translated}`);
+      that.maxVideoBW = translated;
+    } else {
+      that.maxVideoBW = maxVideoBW;
+    }
```
Comment (jcague) [inline erizo_controller/erizoClient/src/Stream.js:87]: should we reapply the encoder parameters here?

File: erizo_controller/erizoClient/src/webrtc-stacks/FirefoxStack.js [hunk 12]
```diff
@@ -3,54 +3,28 @@ import BaseStack from './BaseStack';
...
+      options.streams = [nativeStream];
       const transceiver = that.peerConnection.addTransceiver(track, options);
-      stream.transceivers.push(transceiver);
-      if (track.kind === 'video' && that.simulcast) {
-        getSimulcastParametersForFirefox(transceiver.sender, isScreenshare).catch(() => {});
-      }
+      nativeStream.transceivers.push(transceiver);
+      const parameters = transceiver.sender.getParameters() || {};
+      parameters.encodings = streamInput.generateEncoderParameters();
+      console.warn('parameters to set', parameters);
```
Comment (jcague) [inline erizo_controller/erizoClient/src/webrtc-stacks/FirefoxStack.js:26]: Consider removing this line

**Review by jcague — APPROVED** [review]
LGTM, I only have some minor comments but overall functionality looks right and it's quite promising to improve quality

File: erizo/src/erizo/rtp/BandwidthEstimationHandler.cpp (hunk 2 above)
Comment (lodoyun) [reply erizo/src/erizo/rtp/BandwidthEstimationHandler.cpp:226]: you are right. I kept if for consistency but it does make more sense to have it as a local variable.

File: erizo_controller/erizoClient/src/Stream.js (hunk 11 above)
Comment (lodoyun) [reply erizo_controller/erizoClient/src/Stream.js:87]: We don't always want to apply it immediately. For instance, if you specify a maxVideoBW in `room.publish` we then generate the initial configuration and then include those in the transceiver.

**Pull request merged**